jmespath-core/src/main/java/io/burt/jmespath/function/MergeFunction.java

```diff
@@ -0,0 +1,43 @@
+package io.burt.jmespath.function;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.LinkedHashMap;
+
+import io.burt.jmespath.Adapter;
+
+public class MergeFunction extends JmesPathFunction {
+  public MergeFunction() {
+    super(1, Integer.MAX_VALUE);
+  }
+
+  @Override
+  protected <T> T internalCall(Adapter<T> adapter, List<ExpressionOrValue<T>> arguments) {
+    if (isObjectArray(adapter, arguments)) {
+      Map<String, T> accumulator = new LinkedHashMap<>();
+      for (ExpressionOrValue<T> argument : arguments) {
+        T value = argument.value();
+        for (String property : adapter.getPropertyNames(value)) {
+          accumulator.put(property, adapter.getProperty(value, property));
+        }
```

Comment:
Well, combining with my previous feedback: there could be a `Iterator<Map.Entry<T,T>> getProperties()` that would not expose an interface that would be so much more complicated than the existing one. For incrementally building an object, a new method like `void putObjectProperty(T object, T property, T value)` would obviously be needed, but I'm not sure that would be too problematic. This interface would also be useful for incremental construction of objects elsewhere (when evaluating other JMESPath expressions that yield objects).